readme.md
# Today List | Todo List 🚀️

Small app where you can add your todo list, set a time limit for todo it and add your favorit YouTube music.

If you don't finish your tasks in the set time, your to-do list will be destroyed

## Functionalities (completed)

* User can add tasks to list ✔️
* User can specify the prority of task ✔️
  * High
  * Medium
  * Lower
* User can change task status ✔️
  * Closed
  * Removed
* User can set limit time for todo tasks ✔️
  * Hours
* Can add your favorit YouTube music ✔️
  * Video id
* User can start the timer ✔️
* Basic controls media player ✔️
  * previus
  * play/pause
  * next
* Tasks are kept even if the tab or browser is closed ✔️
  * LocalStorage
* The list its will be destroyed if the time limit is finished ✔️

For people who are learnign HTML5 (html, css3 y js) can comunicate to collaborate with new features.

* contactozurckz@gmail.com
* [Zurckz Facebook Page](https://www.facebook.com/zurckz2018/)
* [Zurckz Personal Facebook](https://www.facebook.com/zurckz.zoe)
* [Zurckz YouTube](https://www.youtube.com/c/Zurckz2015)

## Built With

* HTML5
* CSS3
* JS
* Very little of JQuery 😱

## Versioning

For the versions available, see the [on this repository](https://github.com/NoeCruzMW/today-list-app/blob/master/changelog.md).

## Authors

* **Noé Curz | Zurckz** - *Initial work* -

See also the list of [contributors](https://github.com/your/project/contributors) who participated in this project.

## License

This project is licensed under the MIT License - see the [LICENSE.md](LICENSE.md) file for details


changelog.md
## [0.0.1] - 2020-07-10

### Added

- Form to add tasks.
- Form to add YouTube video by id.
- Task List.
- Time configuration.
- Start Button.
- Automatically destroy task list when time is over.
- Actions on task row for change his status or destroy it.
- Simple player, according youtube videos configured.


js/Storage.js
/**
 * @author Noe Cruz | Zurckz
 *  Status 
 *  0 == not configured
 *  1 == configured
 *  2 == started
 */
class ZStorage {
    constructor(storageName, schema) {
        this.storageName = storageName;
        this.schema = schema;
        this.__read();
    }

    __read() {
        let existSchema = localStorage.getItem(this.storageName);
        let error = 0;

        if (typeof existSchema !== 'undefined') {
            try {
                this.data = JSON.parse(existSchema.toString());
                error = 1;
            } catch (e) {
                error = 0;
            }
        }

        if (error == 0) {
            localStorage.setItem(this.storageName, JSON.stringify(this.schema));
            this.data = this.schema;
        }
    }

    updateAll(schema) {
        localStorage.setItem(storageName, JSON.stringify(schema));
        this.__read();
    }

    update(key, value) {
        this.data[key] = value;
        localStorage.setItem(this.storageName, JSON.stringify(this.data));
        this.__read();
    }

    get(key) {
        return this.data[key];
    }

    get allData() {
        return this.data;
    }
}

js/ZurckzPlayer.js
class ZurckzPlayer {
    constructor(list) {
        this.events$ = new rxjs.Subject();
        this.events = {
            ready: false,
            currentTrackName: '',
            status: -5,
        };
        this.playList = list;
        this.current = 0;
        this.done = false;
        this.isPlaying = false;
        this.ready = false;
        this.__build();
        this.onReadyEvent = new CustomEvent('ready');
    }

    __updateStatus() {
        this.events$.next(this.events);
    }

    __build() {
        let tag = document.createElement('script');
        tag.src = 'https://www.youtube.com/iframe_api';
        let firstScriptTag = document.getElementsByTagName('script')[0];
        firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);
    }

    onYouTubeIframeAPIReady() {
        this.player = new YT.Player('player', {
            height: '0',
            width: '0',
            videoId: this.playList[this.current],
            events: {
                onReady: this.onPlayerReady.bind(this),
                onStateChange: this.onPlayerStateChange.bind(this),
                onError: this.onPlayerError.bind(this),
            },
        });
    }

    get status$() {
        return this.events$;
    }

    get isReady() {
        return this.isReady;
    }

    onPlayerError(e) {
        console.error(e);
        this.next();
    }

    onPlayerReady(event) {
        this.ready = true;
        this.events.ready = true;
        this.__updateStatus();
    }

    onPlayerStateChange(event) {
        console.log('event: ', event);
        this.events.status = event.data;
        this.__updateStatus();
        if (event.data == YT.PlayerState.PLAYING && !this.done) {
            this.done = true;
        }
        switch (event.data) {
            case YT.PlayerState.PLAYING:
                this.isPlaying = true;
                this.events.currentTrackName = this.player.getVideoUrl(); //this.playList[this.current];
                this.__updateStatus();
                break;
            case YT.PlayerState.ENDED:
                this.isPlaying = false;
                this.next();
                break;
            default:
                break;
        }
    }

    back() {
        this.stop();
        if (this.current > 1) {
            this.current--;
        } else {
            this.current = this.playList.length - 1;
        }
        this.load();
    }

    next() {
        this.stop();
        if (this.current < this.playList.length - 1) {
            this.current++;
        } else {
            this.current = 0;
        }
        this.load();
    }

    load() {
        this.player.loadVideoById(this.playList[this.current], 0, 'large');
        setTimeout(() => {
            this.player.playVideo();
            this.isPlaying = true;
        }, 50);
        this.done = false;
    }

    stop() {
        this.player.stopVideo();
        this.isPlaying = false;
    }

    pause() {
        this.player.pauseVideo();
        this.isPlaying = false;
    }

    play() {
        this.player.playVideo();
        this.isPlaying = true;
    }
}

js/ZurckzTodayList.js
class ZurckzTodayList {
    constructor(config) {
        this.tasks$ = new rxjs.Subject();
        this.songs$ = new rxjs.Subject();
        this.tasks = config.tasks || [];
        this.songs = config.songs || [{
            id: 'zurckz-id-music',
            url: 'hHiQ77uQVOk',
        }, ];
        this.listId = config.list;
        this.songListId = config.songList;
        this.priorities = ['high', 'medium', 'low', 'close'];
        this._fillList();
        this._fillSongList();
    }

    _fillList() {
        let htmlList = document.getElementById(this.listId);
        htmlList.innerHTML = '';
        this.tasks.forEach(task => {
            let action = `<button title="Mark as finished" onclick="closeTask(${task.id})"><i class="fa fa-check" aria-hidden="true"></i></button>`;
            if (task.priority == 3) {
                action = `<button title="Remove from the list" onclick="removeTask(${task.id})"><i class="fa fa-times" aria-hidden="true"></i></button>`;
            }
            htmlList.insertAdjacentHTML(
                'beforeend',
                ` <li class="task">
            <div class="task-a task-action">
                ${action}
            </div>
            <div class="task-a task-description ct">
                <p>
                    ${task.description}
                </p>
            </div>
            <div class="task-a task-priority">
                <span class="${this.priorities[task.priority]}"></span>
            </div>
        </li>`
            );
        });
    }

    _fillSongList() {
        let htmlList = document.getElementById(this.songListId);
        htmlList.innerHTML = '';
        let i = 0;
        this.songs.forEach(song => {
            htmlList.insertAdjacentHTML(
                'beforeend',
                `<li class="link" id="song-${song.id}">
                <a href="${song.url}" target="_blank">
                ${song.url}
                </a>
                <button class="" type="button" onclick="deleteSong('song-${song.id}',${song.id})">
                    <i class="fa fa-trash" aria-hidden="true"></i>
                </button>
            </li>`
            );
        });
    }

    addTask(task) {
        console.log(`Task will to add`, task);
        this.tasks.push(task);
        this.tasks$.next(this.tasks);
        this.tasks = this.tasks.sort((ta, tb) => {
            return parseInt(ta.priority) - parseInt(tb.priority);
        });
        this._fillList();
    }

    closeTask(taskId) {
        this.tasks.map(task => {
            if (taskId === task.id) {
                task.priority = 3;
            }
            return task;
        });
        this.tasks$.next(this.tasks);
        this.tasks = this.tasks.sort((ta, tb) => {
            return parseInt(ta.priority) - parseInt(tb.priority);
        });
        this._fillList();
    }

    removeAllTasks() {
        this.tasks = [];
        this.tasks$.next(this.tasks);
        this._fillList();
    }

    addSong(song) {
        this.songs.push(song);
        this.songs$.next(this.songs);
        this._fillSongList();
    }

    removeSong(song) {
        console.log(song);
        this.songs = this.songs.filter(s => s.id != song);
        this.songs$.next(this.songs);
    }

    removeTask(taskId) {
        this.tasks = this.tasks.filter(s => s.id != taskId);
        this.tasks$.next(this.tasks);
        this._fillList();
    }

    get taskList() {
        return this.tasks$;
    }

    get songList$() {
        return this.songs$;
    }
}

//export default ZurckzTodayList;

js/ZurckzCounterDown.js
class ZurckzCounterDown {
    constructor(config) {
        this.events$ = new rxjs.Subject();
        this.events = {
            finished: false,
        };
        this.countDownTime = config.time + 1;
        this.counter = this.countDownTime;
        this.display = config.display;
        this.speed = config.speed;
        this.active = false;
    }
    start() {
        this.interval = setInterval(() => {
            this.countDownTime--;            
            let hours = Math.floor(this.countDownTime / 3600);
            let minutes = Math.floor(this.countDownTime % 3600 / 60);
            let seconds = Math.floor(this.countDownTime % 3600 % 60);
            document.getElementById(
                this.display
            ).innerHTML = `${hours < 10 ? '0' + hours : hours}:${minutes < 10 ? '0' + minutes : minutes}:${seconds < 10 ? '0' + seconds : seconds}`;
            if (this.countDownTime < 0) {
                this.events.finished = true;
                this.events$.next(this.events);
                this.stop();
            }
        }, 1000);
    }

    setTime(time) {
        this.countDownTime = time;
        this.counter = this.countDownTime;
    }

    stop() {
        this.active = false;
        this.countDownTime = 0;
        clearInterval(this.interval);
        document.getElementById(this.display).innerHTML = `00:00:00`;
    }
    reset() {
        this.stop();
        this.countDownTime = this.counter;
    }

    get isActive() {
        return this.active;
    }

    get currentTime() {
        return this.countDownTime;
    }

    get status$() {
        return this.events$;
    }
}

//export default ZurckzCounterDown;

js/today-list.js
// import ZurckzCounterDown from './ZurckzCounterDown.js';
// import ZurckzTodayList from './ZurckzTodayList.js';
let zStorage = new ZStorage('zurckz-today-list-v2', {
    status: 0,
    time: 0,
    tasks: [],
    songs: [],
    lastClosed: null,
    helper: 0,
});

if (zStorage.get('helper') == 0) {
    $('#help-viewer').removeClass('invisible');
    setTimeout(() => {
        $('#help-viewer').addClass('invisible');
        zStorage.update('helper', 1);
    }, 3500);
}

$('#modal').iziModal({
    borderBottom: false,
    headerColor: 'rgba(0,0,0,0)',
});
$('#settingsModal').iziModal({
    borderBottom: false,
    headerColor: 'rgba(0,0,0,0)',
});
$('#songsModal').iziModal({
    borderBottom: false,
    headerColor: 'rgba(0,0,0,0)',
});
$('#successModal').iziModal({
    transitionIn: 'comingIn',
    transitionOut: 'comingOut',
    transitionInOverlay: 'fadeIn',
    transitionOutOverlay: 'fadeOut',
    timeout: 2000,
    timeoutProgressbar: true,
    headerColor: 'rgb(0, 175, 102)',
    timeoutProgressbarColor: 'rgba(255,255,255,0.5)',
    borderBottom: false,
});

const addTaskButton = document.getElementById('fabAdTask');
addTaskButton.onclick = () => showFormTask();

let counter = new ZurckzCounterDown({
    time: zStorage.get('time'),
    display: 'counterDown',
    speed: 1,
});

let zurckzList = new ZurckzTodayList({
    list: 'task-list',
    songList: 'songList',
    tasks: zStorage.get('tasks'),
    songs: zStorage.get('songs'),
});

let zPlayer = new ZurckzPlayer([
    'I9mjKMB-Ipk',
    '0_19gk1Dc4w',
    'x88ic4vjrh4',
    'GpOcEcbycfo',
    'eZ3D8EaEtdI',
]);

counter.status$.subscribe({
    next: v => {
        if (v.finished) {
            destroyTaskList();
        }
    },
});

zPlayer.events$.subscribe({
    next: v => {
        console.log(v.ready);
        if (v.ready == true) {
            $('#player-surface').removeClass('invisible');
        }
        if (v.status == 1 || v.status == -5 || v.status == 2) {
            setLoader(false);
        } else {
            setLoader(true);
        }
        $('#track-name').text(v.currentTrackName);
    },
});

zurckzList.taskList.subscribe({
    next: v => {
        zStorage.update('tasks', v);
    },
});

zurckzList.songList$.subscribe({
    next: v => {
        zStorage.update('songs', v);
    },
});

window.addEventListener('beforeunload', function(e) {
    let time = new Date().getTime();
    zStorage.update('lastClosed', time);
    zStorage.update('time', counter.currentTime);
    // var confirmationMessage = '\o/';
    // (e || window.event).returnValue = confirmationMessage; //Gecko + IE
    // return confirmationMessage; //Webkit, Safari, Chrome
});

function showSettings() {
    $('#settingsModal').iziModal('open');
}

function showSongs() {
    $('#songsModal').iziModal('open');
}

function showFormTask() {
    $('#modal').iziModal('open');
}

function showSuccess(title, subtitle) {
    $('#successModal').iziModal('open', {
        title: title,
        subtitle: subtitle,
    });
}

function cancelAddTask() {
    $('#modal').iziModal('close');
}

function closeModal(modal) {
    $('#' + modal).iziModal('close');
}

function saveTask(e) {
    e.preventDefault();
    zurckzList.addTask({
        description: $('#input-description').val(),
        priority: $('#input-priority').val(),
        status: 'created',
        id: new Date().getTime(),
    });
    $('#input-description').val('');
    showSuccess('Added task', 'Today list!');
    return false;
}

function saveSongs(e) {
    e.preventDefault();
    let time = $('#input-time').val();
    showSuccess('Ok', 'Today list!');
    return false;
}

function setTime(e) {
    e.preventDefault();
    let time = $('#input-time').val();
    counter.setTime(time * 60 * 60);
    zStorage.update('time', time * 60 * 60);
    showSuccess('Ok', 'Today list!');
    return false;
}

function start() {
    if (!counter.isActive) {
        counter.start();
        zStorage.update('status', 2);
    }
}

function validate(url) {
    // let re = /^(https\:\/\/)?(www\.youtube\.com|youtu\.?be)\/.+$/;
    // return re.test(url);
    return true;
}

function addSong() {    
    let url = $('#input-new-link').val();
    if (validate(url)) {
        zurckzList.addSong({
            id: new Date().getTime() + 1,
            url: url,
        });
        $('#input-new-link').val('');
    } else {
        $('#addError').css('display', 'block');
        setInterval(() => {
            $('#addError').css('display', 'none');
        }, 2500);
    }
}

function deleteSong(id, song) {
    console.log('Remove song ' + id + ' ' + song);
    zurckzList.removeSong(song);
    $(`#${id}`).remove();
}

function onYouTubeIframeAPIReady() {
    zPlayer.onYouTubeIframeAPIReady();
}

function setLoader(value) {
    if (value) {
        $('#loader').removeClass('invisible');
        $('#loader').addClass('visible');
        $('#track-name').removeClass('visible');
        $('#track-name').addClass('invisible');
        return;
    }
    $('#track-name').removeClass('invisible');
    $('#track-name').addClass('visible');
    $('#loader').removeClass('visible');
    $('#loader').addClass('invisible');
}

function toggle() {
    setLoader(true);
    if (zPlayer.isPlaying) {
        zPlayer.pause();
        $('#c-play').addClass('show-c');
        $('#c-play').removeClass('hide-c');
        $('#c-pause').removeClass('show-c');
        $('#c-pause').addClass('hide-c');
    } else {
        $('#c-pause').removeClass('show-c');
        $('#c-play').addClass('hide-c');
        $('#c-pause').removeClass('hide-c');
        $('#c-pause').addClass('show-c');
        zPlayer.play();
    }
}

function next() {
    zPlayer.next();
}

function back() {
    zPlayer.back();
}

function closeTask(id) {
    zurckzList.closeTask(id);
}

function removeTask(id) {
    zurckzList.removeTask(id);
}

function destroyTaskList() {
    Swal.fire({
        title: 'Oops!',
        text: 'time is up, the task list will be removed!',
        icon: 'warning',
        showCancelButton: true,
        confirmButtonColor: '#3085d6',
        cancelButtonColor: '#d33',
        confirmButtonText: 'Yes, remove it!',
    }).then(result => {
        if (result.value) {
            zStorage.update('status', 0);
            counter.stop();
            counter.setTime(0);
            zurckzList.removeAllTasks();
            zPlayer.stop();
            Swal.fire(
                'Removed!',
                'You can add new tasks with a new time!',
                'success'
            );
        }
    });
}

switch (zStorage.get('status')) {
    case 2:
        let currentTime = new Date().getTime();
        let lastTime = zStorage.get('lastClosed');
        let diff = (currentTime - lastTime) / 1000;
        counter.setTime(zStorage.get('time') - diff);

        console.log(`
            ${currentTime}
            ${lastTime}
            ${diff}
            ${zStorage.get ('time')}
        `);

        if (counter.currentTime < 0) {
            destroyTaskList();
        } else {
            counter.start();
        }
        break;

    default:
        break;
}